Fix locking when ClassMetadata is unserialized

buffcode opened the pull request:
|      Q       |   A
|------------- | -----------
| Type         | bug
| BC Break     | no
| Fixed issues | #2278

#### Summary

Caching / unserializing ClassMetadata broke locking functionality

buffcode committed:
Fix locking when ClassMetadata is unserialized

Caching / unserializing ClassMetadata broke locking functionality

Fixes #2278

SenseException commented:
@malarzm The website config lists 2.2.x as the maintained current branch and 2.3.x as the upcoming, while the default branch is set to 2.3.x. Which branch should this PR target?

malarzm commented:
This PR shall prevail :) I'll be closing the other ones and review this one shortly

malarzm reviewed (changes requested):
Thanks for the PR! The change on it's own is great however we need to have it covered by a test to prevent regressions in future. At a glance this test here https://github.com/doctrine/mongodb-odm/blob/06a7f2fcd8dbf7c95218f217fbff8601ba26c92f/tests/Doctrine/ODM/MongoDB/Tests/Mapping/ClassMetadataTest.php#L36 looks like a good candidate.

buffcode committed:
Test serialization of lock/version fields

buffcode commented:
@malarzm I have added assertions to the `ClassMetadataTest` for this issue. I've also added assertions for `isVersioned` and `versionField` though they are not affected, I hope thats okay.

buffcode reviewed (commented):
Adressed in a97aab0c4703ad17eac1efaa9bbb66ad844990ae

malarzm reviewed (approved):
That's great, thanks for adding versioning to the test as well

malarzm commented:
Thanks @buffcode!
